### ISSUE TITLE
renderElement could take a doctype keyword argument

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5560]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#5560|
|Type|enhancement|
|Created|2012-03-17 03:03:39Z|
|Branch|https://github.com/twisted/twisted/tree/renderElement-doctype-5560|

The doctype keyword argument would cause renderElement to write out the given doctype before it begins flattening the template and writing that to the request.
```
renderElement(request, element, doctype='<!DOCTYPE html>')
```

is a mild improvement over:

```
request.write('<!DOCTYPE html>')
renderElement(request, element)
```

but it is a major improvement if the doctype attribute has a default value of ```<!DOCTYPE html>```.

<details><summary>Searchable metadata</summary>

```
trac-id__5560 5560
type__enhancement enhancement
reporter__dreid dreid
priority__normal normal
milestone__ 
branch__branches_renderElement_doctype_5560 branches/renderElement-doctype-5560
branch_author__dreid dreid
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1331953419000000 1331953419000000
changetime__1332819959000000 1332819959000000
version__None None
owner__dreid dreid
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_2"></a>|@dreid commented|
|-|-|

(In [33912]) Branching to 'renderElement-doctype-5560'

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_3"></a>|@dreid commented|
|-|-|

(In [33916]) Add a doctype keyword argument which will get written as the first line of the rendered response.

Refs [#5560](https://github.com/twisted/twisted/issues/5560)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Just waiting on some [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/renderElement-doctype-5560).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm set owner to @dreid|
|-|-|

Good change! Builds initiated by glyph pass (minus some Windows failures I think are unrelated). Some comments:

 * Please add a news file.
 * The section on DOCTYPE in the Twisted Web templates howto documentation should be adjusted to point to this new feature, and that you don't have to write it out manually anymore.
 * Reading the current text there, I'm afraid that having a default doctype being written out now, constitutes a non-compatible change. I.e. we've instructed people to write their own doctype. Unmodified, adding a default ourselves will result in two of them being written out.
 * In the docstring of `renderElement`:
   * Please use `C{str}` rather than `C{string}`.
   * Put the default docstring in quotes: `C{'<!DOCTYPE html5>'}`.
   * End the sentence for that parameter with a full-stop.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_7"></a>|@dreid commented|
|-|-|

(In [33984]) Fix up doctype docstring.  Add a news file.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_8"></a>|@dreid commented|
|-|-|

Replying to [ralphm](#note_6):
> Good change! Builds initiated by glyph pass (minus some Windows failures I think are unrelated). Some comments:

Thanks for the review ralphm!

>  * Please add a news file.

Done.

>  * The section on DOCTYPE in the Twisted Web templates howto documentation should be adjusted to point to this new feature, and that you don't have to write it out manually anymore.

There is a separate ticket [#5559](https://github.com/twisted/twisted/issues/5559) about documenting `renderElement` in the template howto.  Clearly the `doctype` parameter should be documented as part of that.

>  * Reading the current text there, I'm afraid that having a default doctype being written out now, constitutes a non-compatible change. I.e. we've instructed people to write their own doctype. Unmodified, adding a default ourselves will result in two of them being written out.

`renderElement` has never been included in a release so this is not a backwards incompatible change.

>  * In the docstring of `renderElement`:
>    * Please use `C{str}` rather than `C{string}`.
>    * Put the default docstring in quotes: `C{'<!DOCTYPE html5>'}`.
>    * End the sentence for that parameter with a full-stop.

Done.

> 
> Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_10"></a>|@ralphm set owner to @dreid|
|-|-|

Replying to [dreid](#note_8):
> Replying to [ralphm](#note_6):
> >  * Please add a news file.
> 
> Done.

Nit: the news entry should be sentence with `renderElement` as the [subject of the sentence](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles).

> >  * The section on DOCTYPE in the Twisted Web templates howto documentation should be adjusted to point to this new feature, and that you don't have to write it out manually anymore.
> 
> There is a separate ticket [#5559](https://github.com/twisted/twisted/issues/5559) about documenting `renderElement` in the template howto.  Clearly the `doctype` parameter should be documented as part of that.

Ok, that's fine.

> >  * Reading the current text there, I'm afraid that having a default doctype being written out now, constitutes a non-compatible change. I.e. we've instructed people to write their own doctype. Unmodified, adding a default ourselves will result in two of them being written out.
> 
> `renderElement` has never been included in a release so this is not a backwards incompatible change.

I didn't notice it was that new. Never mind.

When you have addressed the above nit, please merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_11"></a>|@dreid commented|
|-|-|

(In [34015]) Tweak the grammar of the news file entry.

Refs [#5560](https://github.com/twisted/twisted/issues/5560)

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_12"></a>|@dreid set status to `closed`|
|-|-|

(In [34016]) Merge renderElement-doctype-5560: renderElement handles doctype output.

Author: dreid
Reviewer: ralphm
Fixes: [#5560](https://github.com/twisted/twisted/issues/5560)

Add a doctype keyword argument that will be written as the first line of the
response.  Defaulting to the HTML5 doctype '<!DOCTYPE html>'.
